esp/logging/logginglib/logconfigptree.hpp

```diff
@@ -0,0 +1,111 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2021 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef _LOGCONFIGPTREE_HPP_
+#define _LOGCONFIGPTREE_HPP_
+
+#include "jptree.hpp"
+#include "jstring.hpp"
+#include "tokenserialization.hpp"
+
+namespace LogConfigPTree
+{
+    inline const char* queryProp(const IPTree& node, const char* xpath)
+    {
+        const char* value = nullptr;
+        if (!isEmptyString(xpath))
+        {
+            const char* delim = strrchr(xpath, '/');
+            size_t attrIndex = (delim ? delim - xpath + 1 : 0);
+            if (xpath[attrIndex] != '@')
+            {
+                StringBuffer altXPath(xpath);
+                altXPath.insert(attrIndex, '@');
+                value = node.queryProp(altXPath);
+                if (value)
+                    return value;
+            }
+            value = node.queryProp(xpath);
+        }
+        return value;
+    }
+
+    inline const char* queryProp(const IPTree* node, const char* xpath)
+    {
+        if (node)
+            return queryProp(*node, xpath);
+        return nullptr;
+    }
+
+    inline bool getProp(const IPTree& node, const char* xpath, StringBuffer& value)
+    {
+        const char* raw = queryProp(node, value);
+        if (raw)
+        {
+            value.set(raw);
+            return true;
+        }
+        return false;
+    }
+
+    inline bool getProp(const IPTree* node, const char* xpath, StringBuffer& value)
+    {
+        if (node)
+            return getProp(*node, xpath, value);
+        return false;
+    }
+
+    template <typename value_t, typename default_t>
+    value_t getProp(const IPTree& node, const char* xpath, const default_t& defaultValue)
+    {
+        const char* raw = queryProp(node, xpath);
+        if (raw)
+        {
+            static TokenDeserializer deserializer;
+            value_t value = value_t(defaultValue);
+            if (deserializer(raw, value) == Deserialization_SUCCESS)
+                return value;
+        }
+        return value_t(defaultValue);
```

Comment:
Unless I missed it, this template function doesn't seem to be used, but could there be an issue with using two disparate types for value and default? The will the statement `value_t(defaultValue)` always yield a safe or correct result if default_t can't be cast to value_t or there is no value_t constructor taking a default_t?

Reply:
For invalid default types, such as const char* as a default for int, compiler errors are expected. Unfortunately, compiler warnings are not expected for sign and size mismatches. The questionable cast will be replaced with a function call that attempts to identify improper casts. The casts will happen regardless, but errors and warnings will be recorded for mismatched values.